File: 6_图论算法/leetcode3310.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    void DFS(vector<vector<int>>& graph, vector<bool>& isVisited, int node) {
        if (isVisited[node]) {
            return;
        }
        isVisited[node] = true;
        for (int nextNode : graph[node]) {
            if (!isVisited[nextNode]) {
                DFS(graph, isVisited, nextNode);
            }
        }
    }

    vector<int> remainingMethods(int n, int k, vector<vector<int>>& invocations) {
        // 构建有向图
        vector<vector<int>> graph(n);
        for (const auto& edge : invocations) {
            int x = edge[0];
            int y = edge[1];
            graph[x].push_back(y);
        }

        // 初始化访问标记数组
        vector<bool> isVisited(n, false);
        vector<int> ans;

        // 从可疑节点k开始DFS遍历
        DFS(graph, isVisited, k);

        // 收集未被访问的节点（非可疑方法）
        for (int i = 0; i < n; ++i) {
            if (!isVisited[i]) {
                ans.push_back(i);
            }
        }

        // 检查是否存在剩余方法调用可疑方法的情况
        for (const auto& edge : invocations) {
            int x = edge[0];
            int y = edge[1];
            // 如果x在剩余方法中 且y是可疑方法
            if (!isVisited[x] && isVisited[y]) {
                // 返回所有节点
                vector<int> allNodes(n);
                for (int i = 0; i < n; ++i) {
                    allNodes[i] = i;
                }
                return allNodes;
            }
        }

        return ans;
    }
};

File: 6_图论算法/leetcode3310.py
from typing import List


class Solution:
    def DFS(self, graph: List[List[int]], isVisited: List[int], node: int) -> None:
        if isVisited[node]:
            return
        isVisited[node] = True
        for nextNode in graph[node]:
            if not isVisited[nextNode]:
                self.DFS(graph, isVisited, nextNode)

    def remainingMethods(
        self, n: int, k: int, invocations: List[List[int]]
    ) -> List[int]:
        ##有向无权图
        ##根据题意可以知道，处于k这个联通块之外的 节点即为答案
        graph = [[] for _ in range(n)]
        for x, y in invocations:
            graph[x].append(y)  ##有向图
        isVisited = [False] * n
        ans = []
        self.DFS(graph, isVisited, k)
        """
        你需要删除所有可能有 bug 的方法。
        如果删除后无法编译（剩余的方法调用了可疑的方法），那么返回数组 [0,1,2,⋯,n−1]。
        """
        for i in range(n):
            if not isVisited[i]:
                ans.append(i)
        ## ans是删除了可疑方法的剩余方法。看看是否有节点调用了可疑方法，如果有，那么直接返回原数组
        for x, y in invocations:
            ## 如果x是剩余方法，但是y是可疑方法，那么就返回
            if not isVisited[x] and isVisited[y]:
                return [i for i in range(n)]
        return ans


if __name__ == "__main__":
    n = 3
    k = 2
    invocations = [[1, 2], [0, 1], [2, 0]]
    n = 5
    k = 0
    invocations = [[1, 2], [0, 2], [0, 1], [3, 4]]
    invocations = [[1, 2], [0, 1], [3, 2]]
    invocations = [[1, 2], [0, 2], [0, 1], [3, 4]]
    a = Solution()
    print(a.remainingMethods(5, 0, invocations))
